compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchTensorPass.cpp

```diff
@@ -0,0 +1,221 @@
+
+  return success();
+}
+
+/// Rewrite top-level ExtractSliceOps to FlowSliceOps or wrap them in a
+/// dispatch region.
+LogicalResult convertExtractSliceOps(
+    mlir::TensorDimTrackingRewriter &rewriter, mlir::FunctionOpInterface funcOp,
+    SmallVector<IREE::Flow::DispatchWorkgroupsOp> &workgroupsOps) {
+  // Find eligible ExtractSliceOps.
+  SmallVector<tensor::ExtractSliceOp> extractSliceOps;
+  funcOp.walk([&](tensor::ExtractSliceOp op) {
+    if (!isInDispatchRegion(op))
+      extractSliceOps.push_back(op);
+  });
+
+  // Rewrite ExtractSliceOps to FlowSliceOps.
+  SmallVector<Operation *> remainingExtractSliceOps;
+  for (tensor::ExtractSliceOp extractSliceOp : extractSliceOps) {
+    if (failed(convertExtractSliceOpToFlowSliceOp(rewriter, extractSliceOp))) {
+      remainingExtractSliceOps.push_back(extractSliceOp);
+    }
+  }
+
+  // Create a DispatchWorkgroupsOp for every remaining ExtractSliceOp.
+  FailureOr<SmallVector<IREE::Flow::DispatchWorkgroupsOp>> newWorkgroupsOps =
+      wrapInWorkgroupsOp(rewriter, remainingExtractSliceOps);
+  if (failed(newWorkgroupsOps))
+    return failure();
+  workgroupsOps.append(newWorkgroupsOps->begin(), newWorkgroupsOps->end());
+
+  return success();
+}
+
+namespace {
+struct DispatchTensorPass
+    : public IREE::Flow::impl::DispatchTensorPassBase<DispatchTensorPass> {
+  using IREE::Flow::impl::DispatchTensorPassBase<
+      DispatchTensorPass>::DispatchTensorPassBase;
+  void runOnOperation() override;
+};
+} // namespace
+
+void DispatchTensorPass::runOnOperation() {
+  mlir::FunctionOpInterface funcOp = getOperation();
+  mlir::TensorDimTrackingRewriter rewriter(funcOp);
+  mlir::MLIRContext *context = &getContext();
+
+  auto workgroupsOps = SmallVector<IREE::Flow::DispatchWorkgroupsOp>();
+  funcOp->walk([&](Flow::DispatchWorkgroupsOp workgroupsOp) {
+    workgroupsOps.push_back(workgroupsOp);
+  });
+
+  // Rewrite InsertSliceOps to FlowUpdateOps.
+  if (failed(convertInsertSliceOps(rewriter, funcOp, workgroupsOps))) {
+    funcOp->emitOpError(
+        "failed to create dispatch region for `tensor.insert_slice`");
+    return signalPassFailure();
+  }
+
+  // Rewrite ExtractSliceOps to FlowUpdateOps.
+  if (failed(convertExtractSliceOps(rewriter, funcOp, workgroupsOps))) {
+    funcOp->emitOpError(
+        "failed to create dispatch region for `tensor.extract_slice`");
+    return signalPassFailure();
+  }
+
+  // Canonicalize to flow.tensor ops.
+  RewritePatternSet convertToFlowPatterns(context);
+  IREE::Flow::populateTensorToFlowConversionPatterns(context,
+                                                     convertToFlowPatterns);
+  memref::populateResolveRankedShapedTypeResultDimsPatterns(
+      convertToFlowPatterns);
+  IREE::Flow::TensorReshapeOp::getCanonicalizationPatterns(
+      convertToFlowPatterns, context);
+  IREE::Flow::TensorBitCastOp::getCanonicalizationPatterns(
+      convertToFlowPatterns, context);
```

Comment:
This is new to me. Are we generating tensor bit_cast ops at flow level?

Reply:
I'm not entirely sure, I would assume `tensor.bitcast` ops are lowered to flow and then cleaned up here. The was in the original FormDispatchWorkgroups.cpp, but maybe its not needed?